### ISSUE TITLE
qwen2audioencoder pad test

### ISSUE THREAD

**Issue by LeBeGut** [wuzhiyue111/Codec-Evaluation#39]
def process_sig(self, sig, length):
        """ 
            Process the signal to obtain input_feature and feature_attention_mask
            sig: [B, T]
            length: [B]
            return: tuple(input_feature, feature_attention_mask)
                    input_feature: [B, D, N]
                    feature_attention_mask: [B, N]
        """
        device = sig.device
        batch_size = sig.shape[0] 
        max_length = sig.shape[1]
        sig_lens = (sig.shape[1] * length).long()   # tensor

        max_sample = self.audio_max_length_sec * self.orig_sample_rate
        if torch.any(sig_lens > max_sample):
            raise ValueError(f"Audio length exceeds the maximum allowed length of {self.audio_max_length_sec} seconds.")
        
        # process audio and get audio list
        raw_audio_list = []
        for i in range(batch_size):
            valid_length = min(sig_lens[i].item(), max_length)
            audio_sample = sig[i, :valid_length].cpu().numpy()
            raw_audio_list.append(audio_sample)

        features = self.feature_extractor(
            raw_audio_list, 
            sampling_rate=self.orig_sample_rate, 
            return_tensors="pt", 
            padding=True
        )
        
        """
            Convert the maximum duration of the audio (in seconds) 
            to the number of time steps of the Mel-spectrogram.
            mel_hop_length = 160, so the number of time steps of the Mel-spectrogram is:
            1/ (160 / orig_sample_rate) = 1 / 0.01 = 100 
        """
        input_feature = features.input_features.to(device)
        valid_feature_length = self.audio_max_length_sec * 100
        batch_size, _, max_mel_seq_len = input_feature.shape
        if max_mel_seq_len < valid_feature_length:
            pad_features = torch.zeros((batch_size, _, valid_feature_length),
                                      dtype=input_feature.dtype, 
                                      device=device)
            pad_features[:, :, :max_mel_seq_len] = input_feature
            input_features = pad_features
            del pad_features
        
        # Convert original sampling rate to feature rate
        raw_audio_mask = features.attention_mask.to(device)
        valid_mask_length = self.audio_max_length_sec * self.orig_sample_rate
        batch_size = raw_audio_mask.shape[0]
        if raw_audio_mask.shape[1] < valid_mask_length:
            pad_mask = torch.zeros((batch_size, valid_mask_length),
                                   dtype=raw_audio_mask.dtype,
                                   device=device)
            pad_mask[:, :raw_audio_mask.shape[1]] = raw_audio_mask
            raw_audio_mask = pad_mask
            del pad_mask
        
        # Calculate mel spectrogram attention mask with downsampling
        feature_attention_mask = F.max_pool1d(
            raw_audio_mask.unsqueeze(1).float(),    # shape [B, 1, T]
            kernel_size=self.mel_hop_length,        # pooling kernel size
            stride=self.mel_hop_length,             # pooling stride
        ).squeeze(1)

        # boolean mask
        feature_attention_mask = feature_attention_mask > 0

        # Calculate encoder output attention mask dimensions
        batch_size, _, max_mel_seq_len = input_features.shape
        feature_lengths = feature_attention_mask.sum(dim=-1)    # [B]
        encoder_output_lengths = ((feature_lengths - 2) // 2 + 1).long()    # [B] accounting for downsampling
    
        max_length = ((max_mel_seq_len - 2) // 2 + 1)
        attention_mask = torch.ones((batch_size, 1, max_length, max_length), device = device)
        
        for i, len in enumerate(encoder_output_lengths):
            len_val = len.item() if torch.is_tensor(len) else int(len)
            attention_mask[i, :, len_val:, :] = float('-inf')
            attention_mask[i, :, :, len_val:] = float('-inf')

        return input_features, attention_mask
